<center><b>Assignment for Business Intelligence Summer 2021 Internship</b></center>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import datetime as dt

In [2]:
df1=pd.read_csv('BI-Intern first_purchases.csv',date_parser=True)
df2=pd.read_csv('BI-Intern purchases.csv',date_parser=True)

In [3]:
df1

User First Purchase Month First Purchase Product Line  \
0                      1.5.2020                  Restaurant   
1                      1.5.2020                  Restaurant   
2                      1.5.2020                  Restaurant   
3                      1.5.2020                  Restaurant   
4                      1.5.2020                  Restaurant   
...                         ...                         ...   
71252                 1.10.2020                  Restaurant   
71253                 1.10.2020                  Restaurant   
71254                 1.10.2020                  Restaurant   
71255                 1.10.2020                  Restaurant   
71256                 1.10.2020                  Restaurant   

                        User ID               Purchase ID  \
0      57b48aece14f6d64522f8c5f  5ec6b23c96741b6db6266639   
1      59f4803cdf196473b7fc124d  5ec6be1e712a4248e67d2bf0   
2      5a427fbbea63d5000aff0ade  5ec7f07674bbff7edcb5be0d   
3      5a8c70b58fac51000b2ca9d8  5eac48651db7a19612824809   
4      5a9efa9bab3442000c2ba457  5eca9ed670a4dafe5de07698   
...                         ...                       ...   
71252  5f874372e71892d52313343f  5f8745bfae557239f1cbe4db   
71253  5f874b92a026a2143bab6e63  5f874fd5e86004c65a133d47   
71254  5f874e647a7169ef6a023235  5f875660afcc5ba542f3f332   
71255  5f875552c7b1a240baf4cc97  5f87582ed5481f4e72e02218   
71256  5f8755d29765e83cf2c7b52a  5f8757217a5f5d289a90b84a   

                       Venue ID  
0      5d7a32a73b546fccd88a05e9  
1      5cd16542c7420059cc10b792  
2      5e7bce9ae360062306b4a233  
3      5e6ce90a7455f89bbb00be37  
4      5ea18277857aaa1b337ee201  
...                         ...  
71252  5f2d6542365ed912d6c545f2  
71253  5e6ccb03b1d0a9394d072fcd  
71254  5e77a757c55bc12714975c4d  
71255  5f4cfc839232473565cfd423  
71256  5e7b8de7a7a5e36f91c7b17b  

[71257 rows x 5 columns]

In [4]:
df1.count()

User First Purchase Month      71257
First Purchase Product Line    71257
User ID                        71257
Purchase ID                    71257
Venue ID                       71257
dtype: int64

In [5]:
df2

Purchases Time Delivered  Product line                   User ID  \
0                      1.8.2020    Restaurant  55080d6585f6661a3b37f7a5   
1                      1.8.2020    Restaurant  55080d6585f6661a3b37f7a5   
2                      1.7.2020  Retail store  556202537f84a43ac29ec7cb   
3                      1.7.2020  Retail store  556202537f84a43ac29ec7cb   
4                      1.5.2020    Restaurant  57b48aece14f6d64522f8c5f   
...                         ...           ...                       ...   
298709                1.10.2020    Restaurant  5f874372e71892d52313343f   
298710                1.10.2020    Restaurant  5f874b92a026a2143bab6e63   
298711                1.10.2020    Restaurant  5f874e647a7169ef6a023235   
298712                1.10.2020    Restaurant  5f875552c7b1a240baf4cc97   
298713                1.10.2020    Restaurant  5f8755d29765e83cf2c7b52a   

                     Purchase ID                  Venue ID  
0       5f3d81ef6cce0eeb05225f98  5e8b2c8be050b99f26647804  
1       5f3c11f5a90667e766b04c19  5e8b2c8be050b99f26647804  
2       5f0aec594c3cbeb4049ea36a  5ec2ab4401f812138e435c89  
3       5f184a3c88046da058c5ed1d  5ec2ab4401f812138e435c89  
4       5ecf98f34d3d68dc996dad7a  5cb4425ae93cce000d20612f  
...                          ...                       ...  
298709  5f8745bfae557239f1cbe4db  5f2d6542365ed912d6c545f2  
298710  5f874fd5e86004c65a133d47  5e6ccb03b1d0a9394d072fcd  
298711  5f875660afcc5ba542f3f332  5e77a757c55bc12714975c4d  
298712  5f87582ed5481f4e72e02218  5f4cfc839232473565cfd423  
298713  5f8757217a5f5d289a90b84a  5e7b8de7a7a5e36f91c7b17b  

[298714 rows x 5 columns]

In [6]:
df2.count()

Purchases Time Delivered    298714
Product line                298714
User ID                     298714
Purchase ID                 298714
Venue ID                    298714
dtype: int64

In [7]:
# Change type of column to datetime
df2['Purchases Time Delivered'] = pd.to_datetime(df2['Purchases Time Delivered'], format="%d.%m.%Y")

In [8]:
df2.dtypes

Purchases Time Delivered    datetime64[ns]
Product line                        object
User ID                             object
Purchase ID                         object
Venue ID                            object
dtype: object

## Searching for top online food ordering:

In [51]:
len(df2['Venue ID'].unique())

1503

In [52]:
# Top selling units available on service
top=df2['Venue ID'].value_counts()
top_extract=pd.DataFrame({'Venue ID':top.index,'Count':top.values})

In [53]:
top_extract['Count'].describe()

count    1503.000000
mean      198.745176
std       371.440041
min         1.000000
25%        20.000000
50%        74.000000
75%       215.500000
max      4602.000000
Name: Count, dtype: float64

In [54]:
extract=df2[['Product line','Venue ID']]

In [55]:
top_sell_units=pd.merge(top_sell_unit,extract,how='inner',on='Venue ID')

In [56]:
top_sell_units=top_sell_units.drop_duplicates(keep='first')

In [57]:
# Top 100 selling units
top_sell_units.head(20)

Venue ID  Count  Product line
0      5e77a370fee9ef1ba62d6454   4602    Restaurant
4602   5e7dc5c6d12956517345e7d0   4235    Restaurant
8837   5e74b3973f3b342b1112c795   4176    Restaurant
13013  5bbc6b0b9a23e8000c10ce2c   3250    Restaurant
16263  5e7cf0de1d5f2d9994b536ba   3172    Restaurant
19435  5e7747d9896d723f8302370d   2708    Restaurant
22143  5e54e9202149d59d4f476298   2363    Restaurant
24506  5e73cb5d891277fa4aa54e24   2162  Retail store
26668  5e7781b4f4694683def749a6   2076    Restaurant
28744  5e6ceab07df646d28037d17d   1871    Restaurant
30615  5e74b46d8829822e4229c7f8   1824    Restaurant
32439  5e7b8de7a7a5e36f91c7b17b   1706    Restaurant
34145  5e77a757c55bc12714975c4d   1653    Restaurant
35798  5e7663ebe3845b5fffaff7db   1634    Restaurant
37432  5e74af9eb90fb76d88cca48a   1622    Restaurant
39054  5bc066923bb2a0000b11c6d7   1590    Restaurant
40644  5e7b8cb0d43edea68c4ff05e   1566    Restaurant
42210  5d95eebb1ac8e2ab9bab84e1   1565    Restaurant
43775  5e777d49b9e3054b13f35214   1549    Restaurant
45324  5e6bd3a31cafc5bb101639ae   1539    Restaurant

Most of orders come from restaurants field

## Extracting table with those lost user data from first_purchases:

In [9]:
len(df2['User ID'].unique())

74568

In [10]:
len(df1['User ID'].unique())

71257

With the above results, we are able to recognize conficting data between the amount of customers in df1 (first_purchase table) and df2 (purchases table), specifically 71257 and 74568. It means there are 3311 customers that have purchases listed in df2 not in df1.

In [11]:
a=list(df1['User ID'].unique())

In [12]:
b=list(df2['User ID'].unique())

In [13]:
missing_data=[]
for i in b:
    if i not in a:
        missing_data.append(i)
missing_data

['556202537f84a43ac29ec7cb',
 '57b5a1c7791b5a4aafe403ef',
 '5840920d79f65102d70ede18',
 '58653b9279f65127b206770e',
 '58d7fba58033271c5a3a9caf',
 '599ee35379f65143c0b3c94d',
 '5a098af562f55f06c26eac91',
 '5a82dc2e615ca5000a5fb303',
 '5a8ffa198fac510009315335',
 '5a968a9c572fd6000a09baee',
 '5ad5dee2826817000a64d554',
 '5adce6095400c6000a3e4fed',
 '5ae1e21d5400c6000b77f8d3',
 '5ae4adaf919fae000ccfd99a',
 '5af6f689986b9f000ced8828',
 '5afd4f8b820567000aea5619',
 '5b0017f4a7a356000c65b46b',
 '5b027bb3820567000aeaa9f9',
 '5b03067ba7a356000c65e89f',
 '5b03e74fa7a356000c65f0d5',
 '5b044051820567000aeacdd7',
 '5b084bc38be1dd000c359495',
 '5b0957928be1dd000c35a3f8',
 '5b0d0cd98be1dd000c35ee31',
 '5b0d4fb58be1dd000c35f7e7',
 '5b0d59e352949b000af60b70',
 '5b0d637452949b00098c4c28',
 '5b0d83498be1dd000c36002a',
 '5b0d8c8a8be1dd000c3601d9',
 '5b0e2ef98be1dd000ab5283e',
 '5b0e7cfe52949b000af61c1b',
 '5b0e82318be1dd000ab52d99',
 '5b1159d8cacdac000b49753f',
 '5b11bee552949b00098cb296',
 '5b1291ff5294

In [14]:
len(missing_data)

3311

In [15]:
missing_data=pd.DataFrame(missing_data,columns=['User ID'])

In [16]:
missing_data

User ID
0     556202537f84a43ac29ec7cb
1     57b5a1c7791b5a4aafe403ef
2     5840920d79f65102d70ede18
3     58653b9279f65127b206770e
4     58d7fba58033271c5a3a9caf
...                        ...
3306  5f762fb1fd6b85d071d65228
3307  5f7849898c0a1d38c21b53b1
3308  5f7eea19b461d6ab65fcde83
3309  5f80d6e62792b18d4381dad2
3310  5f842675751eb1dd8b080c47

[3311 rows x 1 columns]

In [17]:
df3=pd.merge(df2,missing_data,how='inner',on=['User ID'])

In [18]:
# Table shows the missing data purchases of users in df2 not in df1
df3

Purchases Time Delivered  Product line                   User ID  \
0                   2020-07-01  Retail store  556202537f84a43ac29ec7cb   
1                   2020-07-01  Retail store  556202537f84a43ac29ec7cb   
2                   2020-10-01    Restaurant  57b5a1c7791b5a4aafe403ef   
3                   2020-10-01    Restaurant  57b5a1c7791b5a4aafe403ef   
4                   2020-09-01    Restaurant  5840920d79f65102d70ede18   
...                        ...           ...                       ...   
10140               2020-10-01    Restaurant  5f7849898c0a1d38c21b53b1   
10141               2020-10-01    Restaurant  5f7eea19b461d6ab65fcde83   
10142               2020-10-01    Restaurant  5f80d6e62792b18d4381dad2   
10143               2020-10-01    Restaurant  5f80d6e62792b18d4381dad2   
10144               2020-10-01    Restaurant  5f842675751eb1dd8b080c47   

                    Purchase ID                  Venue ID  
0      5f0aec594c3cbeb4049ea36a  5ec2ab4401f812138e435c89  
1      5f184a3c88046da058c5ed1d  5ec2ab4401f812138e435c89  
2      5f85cf115a38e7d584abeff6  5e779ec5c96b9193eb284c14  
3      5f79fb5a06c145cdc12d09e0  5eec90cb5f3401e081fa2ae1  
4      5f63240a2c027f3009ff1e75  5f3a8ce44d768c7c157cce76  
...                         ...                       ...  
10140  5f784c7fb5453f963c50507f  5d95f08bd205a22b10440290  
10141  5f7ef84783f13ce49c005482  5e7b8cb0d43edea68c4ff05e  
10142  5f815630521757b130d2309d  5f61c72abea978646e490d98  
10143  5f81a27b7b3991fc47afca9f  5f16add1e1e64c3df0d7dc34  
10144  5f85783a940b2c0afb15a351  5e981da56579a4111bc203ce  

[10145 rows x 5 columns]

## Exploring statistics on purchases from each customers:

In [24]:
# Table to explore purchase count from each customers
sample=df2['User ID'].value_counts()
retention=pd.DataFrame({'User ID':sample.index, 'Purchase Count':sample.values})

In [ ]:
retention

In [ ]:
# Top 20 customers with the most regular usages
retention.head(20)

In [ ]:
# Get insights from collective purchase count
retention['Purchase Count'].describe()

In [ ]:
# Using boxplot to get statistics
plt.figure(figsize=(8,3))
sns.boxplot(retention['Purchase Count'], sym='')
plt.xticks(fontsize=20)
plt.xlabel('Purchase Count',fontsize=20)
sns.despine(left=True,bottom=True)

Median purchase count is 2 purchases<br>
75% of the users span from 1 to 4 purchases during a period

In [ ]:
# Categorizing customers usage
once=0
twice_to_tenth=0
rarely=0
regular=0
for j in retention['Purchase Count']:
    if j == 1:
        once+=1
    elif j > 1 and j <=10:
        twice_to_tenth+=1
    elif j > 10 and j <= 24:
        rarely+=1
    elif j > 20:
        regular+=1

In [ ]:
print(once,twice_to_tenth,rarely,regular)

In [ ]:
plt.figure(figsize=(12,10))
sizes=[once,twice_to_tenth,rarely,regular]
labels=('once','twice_to_tenth','rarely','regular')
y_pos=np.arange(len(labels))
ax=plt.barh(y_pos,sizes, align='center', alpha=0.7,color='#008B8B')
plt.yticks(y_pos,labels,fontsize=20)
plt.xticks(fontsize=20)
plt.ylabel('Customer Retention',fontsize=20)
plt.xlabel('Count',fontsize=20)
plt.title('Types of Customer',fontsize=20)
sns.despine(left=True,bottom=True)
plt.show()

We can see that, most of customers used the service only once, next is a group of customers who purchased from twice to tenth
in the given period time <br>
Rarely customers class is in a puchasing range between more than ten times and average once a week in 24 weeks, individually 4392. <br>
Regular customers are the least group <br>
It could be due to the delivery fee ( based on distance), somehow delaying in delivery, less discount season from restaurants<br> or retail stores together with unappropriated food price have led to the difficulty to keep customers

## Data preparation for cohort analysis:

In [ ]:
# Using datetime library, we extracted the month, year
df2['month'] = df2['Purchases Time Delivered'].dt.month
df2['year']=df2['Purchases Time Delivered'].dt.year

In [ ]:
# In order to sort them in order, we should multiply the year by 100 and add the month to that value. 
# This will create an order that can be sorted correctly.
df2['cohort'] = df2.apply(lambda row: (row['year'] * 100) + (row['month']), axis=1)

In [ ]:
df2

In [ ]:
b=df2['month'].value_counts().sort_index()

In [ ]:
amount=pd.DataFrame({'Month':b.index,'Purchase_Count':b.values})

In [ ]:
amount

In [ ]:
sns.set_style('white')
plt.figure(figsize=(12,10))
objects = amount['Month']
y_pos = np.arange(len(objects))
performance = amount['Purchase_Count']
plt.bar(y_pos, performance, align='center',color='#483D8B',alpha=0.9)
plt.xticks(y_pos, objects,fontsize=20)
plt.yticks(fontsize=20)
plt.ylabel('Purchase_Count',fontsize=20)
plt.xlabel('Monthly',fontsize=20)
plt.title('Amount of Usage',fontsize=20)
sns.despine(left=True, bottom=True)
plt.show()

Mostly orders increase from August and September, specifically reach a peak in September with 81162 purchases.

In [ ]:
# After we create a column called ‘cohort’, we will group all the orders by the CustomerID 
# and find the minimum cohort to find the cohort they belong in as their first purchase.
cohorts = df2.groupby('User ID')['cohort'].min().reset_index()

In [ ]:
cohorts

In [ ]:
# We’ll rename the columns to CustomerID and first cohort 
# Merge new DataFrame with our previous data
cohorts.columns = ['User ID', 'first_cohort']
df2 = df2.merge(cohorts, on='User ID', how='left')

In [ ]:
# Define types of product line
df2['Product line'].unique()

In [ ]:
# Call the “value_counts” function on cohorts column and reset the index
headers = df2['cohort'].value_counts().reset_index()

In [ ]:
headers.columns = ['Cohorts', 'Count']

In [ ]:
# Result in two columns of all cohorts and the amount occurs
headers

In [ ]:
headers = headers.sort_values(['Cohorts'])['Cohorts'].to_list()

In [ ]:
# Result in an ordered list of cohorts
headers

In [ ]:
# In order to create a cohort analysis, we need to create a DataFrame that has an index of each user’s first month of making a purchase 
# and the amount of times that the percent that made a purchase in the subsequent months.
# Cohort distance will be the amount of months between the current order and the first purchase from the customer
df2['cohort_distance'] = df2.apply(lambda row: (headers.index(row['cohort']) - headers.index(row['first_cohort'])) if (row['first_cohort'] != 0 and row['cohort'] != 0)else np.nan, axis=1)

In [ ]:
df2

## Cohort based monthly retention for Restaurant product line:

In [ ]:
# Product line for restaurant
restaurant=df2[df2['Product line']=='Restaurant']

In [ ]:
restaurant

There are 287921 orders from Restaurant

In [ ]:
restaurant['cohort_distance'].unique()

In [ ]:
# Create a DataFrame that has its rows mark the first purchase cohort while the columns are the months since. 
# The values inside will signify how many customers from that cohort purchases in that period. 
# The first column will always be the largest.
cohort_pivot_restaurant = pd.pivot_table(restaurant, index='first_cohort', columns='cohort_distance', values='User ID', aggfunc=pd.Series.nunique)

In [ ]:
# Pivot_table method allows us to pivot our data so that we can calculate the amount of unique customers 
# in each cohort based on that purchases distance from their first purchase in months.
cohort_pivot_restaurant

In [ ]:
# Divide each row by the first column 
# so that we can have a percentage of the customers that have returned from that cohort to make a purchase
# by using the div function
cohort_pivot_restaurant = cohort_pivot_restaurant.div(cohort_pivot_restaurant[0],axis=0)

In [ ]:
cohort_pivot_restaurant

In [ ]:
# Heatmap out of cohort analysis
plt.figure(figsize=(12,10))
sns.set(font_scale=2)
y_labels = [str(int(header)%100) + '-' + str(int(header)/100) for header in headers]
x_labels = range(0, len(y_labels))
plt.yticks(ticks=headers, labels=y_labels)
plt.xticks(ticks=x_labels, labels=x_labels)
plt.title("Cohort Analysis Restaurant")
sns.heatmap(cohort_pivot_restaurant, annot=True, fmt='.0%', mask=cohort_pivot_restaurant.isnull(), square=True, linewidths=.5, cmap=sns.cubehelix_palette(8),robust=True)
plt.xlabel('Months After First Purchase')
plt.ylabel('First Purchase Cohort')
plt.show()

In the restaurant field, all types of customer tend to order less in the next few months. <br>
I assume that the price for food from restaurants are more expensive and maybe less discount season.<br>
Another reason would be shipping fee due to a further location from restaurant to customer's address

## Cohort based monthly retention for Retail product line:

In [ ]:
# Product line for retail store
retail=df2[df2['Product line']=='Retail store']

In [ ]:
retail

There are 10793 orders from Retail Store

In [ ]:
retail['cohort_distance'].unique()

In [ ]:
cohort_pivot_retail = pd.pivot_table(retail, index='first_cohort', columns='cohort_distance', values='User ID', aggfunc=pd.Series.nunique)

In [ ]:
cohort_pivot_retail

In [ ]:
cohort_pivot_retail = cohort_pivot_retail.div(cohort_pivot_retail[0],axis=0)

In [ ]:
cohort_pivot_retail

In [ ]:
plt.figure(figsize=(12,10))
sns.set(font_scale=2)
y_labels = [str(int(header)%100) + '-' + str(int(header)/100) for header in headers]
x_labels = range(5, 11)
plt.yticks(ticks=headers, labels=y_labels)
plt.xticks(ticks=x_labels, labels=x_labels)
plt.title("Cohort Analysis Retail")
sns.heatmap(cohort_pivot_retail, annot=True, fmt='.0%', mask=cohort_pivot_retail.isnull(), square=True, linewidths=.5, cmap=sns.cubehelix_palette(8),robust=True)
plt.xlabel('Months After First Purchase')
plt.ylabel('First Purchase Cohort')
plt.show()

The customer's purchase behavior in the retail ordering is quite fluctuated.<br>
For customers who have first purchases in May tend to order less from next few months<br>
But customers who have first purchases in June, they increasingly order more in August<br>
Customers who have first purchases in July maintain the order rate from August to September (31%) then bottom down in October.<br>
Customers who have first purchases in August have the most order maintaining rate on September (37%)<br>
In conclusion, most customers have registered to the service and regularly used to order food from June to September (Summer and Autumn season)

Beside the retention, I would chose to explore time delivery (on time delivering is important to a customer service), ratings and comments to analyze customers satisfaction to a product/service and listen to their needs to improve more.